### PR TITLE
opam: 2.1.5 -> 2.2.0

### PR DESCRIPTION
**Pull request by kit-ty-kate**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jackyliu16** [comment]
Result of `nixpkgs-review pr 324381` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages failed to build:</summary>
  <ul>
    <li>dune-release</li>
    <li>ocamlPackages.opam-core</li>
    <li>ocamlPackages.opam-format</li>
    <li>ocamlPackages.opam-repository</li>
    <li>ocamlPackages.opam-state</li>
    <li>opam</li>
    <li>opam-installer</li>
    <li>opam-publish</li>
    <li>opam.installer</li>
    <li>opam2json</li>
  </ul>
</details>

**Comment by OlivierNicole** [comment]
Replacing `opam-shebangs.patch` with this https://gist.github.com/OlivierNicole/84b7b764a6a3d32599d5fbeb5ac59d9f seems to patch properly

**Review by OlivierNicole — APPROVED** [review]
The new package definition seems sensible and it bulids and works as expected for me.

File: pkgs/development/tools/ocaml/opam/opam-shebangs.patch [hunk 1]
```diff
@@ -128,4 +132,4 @@ index 00000000..3ea84e2d
 +    stopNest
 +}
 +
-+patchShebangs .
++patchShebangs .
```
[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply pkgs/development/tools/ocaml/opam/opam-shebangs.patch]: fixed

File: pkgs/development/tools/ocaml/opam/default.nix [hunk 2]
```diff
@@ -2,82 +2,18 @@
   ocaml, unzip, ncurses, curl, bubblewrap, Foundation
 }:
 
-assert lib.versionAtLeast ocaml.version "4.02.3";
+assert lib.versionAtLeast ocaml.version "4.08.0";
 
 let
   srcs = {
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ocaml/opam/default.nix:8]: We might as well remove now the entire let in

File: pkgs/development/tools/ocaml/opam/default.nix [hunk 3]
```diff
@@ -88,33 +24,13 @@ in stdenv.mkDerivation {
 
   src = srcs.opam;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ocaml/opam/default.nix:25]:
```suggestion
  src = fetchurl {
    url = "https://github.com/ocaml/opam/releases/download/2.2.0/opam-full-2.2.0-2.tar.gz";
    hash = "sha256-RZ7WTmZD8FxndWOgAOO6oFx2zlKAZOnLnObbSf/zfJc=";
  };
```

Comment (kit-ty-kate) [reply pkgs/development/tools/ocaml/opam/default.nix]: done

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by vbgl** [comment]
See #329608

**Pull request merged**